catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/ResultIterable.java

```diff
@@ -48,53 +52,117 @@
+            QueryRequest queryRequest) {
+        return new ResultIterable(queryFunction, queryRequest, 0);
+    }
+
+    /**
+     * Creates an iterable that will call the {@link CatalogFramework} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param catalogFramework reference to the {@link CatalogFramework} to call to retrieve the
+     *                         results.
+     * @param queryRequest     request used to retrieve the results.
+     * @param maxResultCount   a positive integer indicating the maximum number of results in
+     *                         total to query for
+     */
+    public static ResultIterable resultIterable(CatalogFramework catalogFramework,
+            QueryRequest queryRequest, int maxResultCount) {
+        notNull(catalogFramework, "CatalogFramework cannot be null");
+        isTrue(maxResultCount > 0, "Max Results must be a positive integer", maxResultCount);
+        return new ResultIterable(catalogFramework, queryRequest, maxResultCount);
+    }
+
+    /**
+     * Creates an iterable that will call a {@link QueryFunction} to retrieve the results
+     * that match the {@link QueryRequest} provided.
+     *
+     * @param queryFunction  reference to the {@link QueryFunction} to call to retrieve the
+     *                       results.
+     * @param queryRequest   request used to retrieve the results.
+     * @param maxResultCount a positive integer indicating the maximum number of results in
+     *                       total to query for
+     */
+    public static ResultIterable resultIterable(QueryFunction queryFunction,
+            QueryRequest queryRequest, int maxResultCount) {
+        isTrue(maxResultCount > 0, "Max Results must be zero or positive", maxResultCount);
+        return new ResultIterable(queryFunction, queryRequest, maxResultCount);
+    }
+
+
+    private ResultIterable(CatalogFramework catalogFramework, QueryRequest queryRequest,
+            int maxResultCount) {
+        this(catalogFramework::query, queryRequest, maxResultCount);
+    }
+
+    private ResultIterable(QueryFunction queryFunction, QueryRequest queryRequest,
+            int maxResultCount) {
         notNull(queryFunction, "Query function cannot be null");
         notNull(queryRequest, "Query request cannot be null");
+        isTrue(maxResultCount >= 0, "Max Results cannot be negative", maxResultCount);
 
         this.queryFunction = queryFunction;
         this.queryRequest = queryRequest;
+        this.maxResultCount = maxResultCount;
     }
 
     @Override
     public Iterator<Result> iterator() {
-        return new QueryResultIterator(queryFunction, queryRequest);
+        if (maxResultCount > 0) {
+            return limit(new ResultIterator(queryFunction, queryRequest), maxResultCount);
+        }
+        return new ResultIterator(queryFunction, queryRequest);
+    }
+
+    /**
+     * @param pageSize the maximum page size for each list segment
```

Comment:
❗️ I think I know what's bugging me with having this method here instead of having a top level class. Even though it reduces the number of top-level classes, it feels like it goes against the Single Responsibility Pattern and will make it more difficult for people that don't know the code where to find this paging iterator. If I'm looking for something to return me batches of `Result` objects, I'm not sure I would think to look for a method on this class. However, if I search for `ResultIterator`, both classes would pop-up.

I asked @paouelle's opinion and he also felt that this method didn't belong here. His suggestion was instead to make the `ResultListIterable` more reusable and treat it like a utility class that returns batches of objects of a generic type.

I took a quick look at Guava's `Iterables.partition()` and it looks like we should be able to use that instead of writing our own batch iterator class. Based on their documentation (https://github.com/google/guava/wiki/CollectionUtilitiesExplained#iterables), most `Iterables` are lazy and will not advance until needed, which means that if we were to give our `ResultIterable` to `partition()`, it should only fetch the next batch of results when `next()` is called on the returned `Iterator`. Which is confirmed by looking at the implementation: https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Iterators.java#L620-L649.

So my recommendation would be to get rid of all the `ResultListIterable` and all the code related to batching and add a note telling people to use a utility such as `Iterables.partition()` to do batching.